### PR TITLE
timed: get time and timezone from modem

### LINKED ISSUE #30702: [$300 bounty] timed: get time and timezone from the modem
The comma 3/3X has a [Quectel EG25-G ](https://www.quectel.com/product/lte-eg25-g), and we should be able to pull the time and timezone from the cell network. If this ends up working well, this will remove the timezoned daemon, a bunch of panda RTC code, and dependence on an RTC battery.

The goal is to write a small Python daemon to get the current time and timezone. Requirements:
* works without a SIM, an inactive SIM, and an active SIM
* must replace the `timezoned` functionality
* must be fast (aside from modem bootup and ModemManager startup time)
* works with ModemManager running
* needs a unit test, [this](https://github.com/commaai/openpilot/blob/master/system/qcomgpsd/tests/test_qcomgpsd.py) is a good example of one already using the modem

### PR DESCRIPTION
**Pull request by royjr**
may fix #30702
```
[160236.432168] process.py:40 - : Setting time based on modem
[160236.748426] timed.py:85 - : Setting timezone to Etc/GMT+5
[160236.942801] timed.py:88 - : Setting time to 2023-12-30 04:21:27
```
```
comma@comma-bc60153f:/data/openpilot$ date
Fri 29 Dec 2023 11:21:33 PM -05
```

### CODE REVIEW

File: system/timed.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  return datetime.strptime(modem_output[19:-8], "%Y/%m/%d,%H:%M:%S").strftime("%Y-%m-%d %H:%M:%S")
+
+
+def main() -> NoReturn:
+  while True:
+    time.sleep(60)
+
+    is_onroad = not params.get_bool("IsOffroad")
+    if is_onroad:
+      continue
```
Comment (adeebshihadeh) [inline system/timed.py:74]: remove this, we want to always update it

File: release/files_tici [hunk 2]
```diff
@@ -4,7 +4,7 @@ third_party/snpe/aarch64-ubuntu-gcc7.5/*
 third_party/mapbox-gl-native-qt/include/*
 third_party/acados/larch64/**
 
-system/timezoned.py
```
Comment (adeebshihadeh) [inline release/files_tici:7]: revert the changes to timezoned. for this PR, I just want to add timed but not run it yet

**Review by adeebshihadeh — CHANGES_REQUESTED** [review]

**Review by adeebshihadeh — COMMENTED** [review]
Nice, this is looking pretty good. The last thing we need is an integration test that sets up each state we'll need to test (active SIM, missing SIM, and inactive SIM), runs `timed` for a cycle, then checks the time was set properly.

I can't immediately think of a way to test "inactive", but active and missing should be easy to test for with just AT commands and an active SIM.

**Comment by royjr** [comment]
Tried a few different options that didn't work

#### AT+CFUN
```
comma@comma-bc60153f:/data/openpilot/system$ mmcli -m 0 --command AT+CFUN=0
response: ''
comma@comma-bc60153f:/data/openpilot/system$ mmcli -m 0 --command AT+QLTS=1
error: command failed: 'GDBus.Error:org.freedesktop.ModemManager1.Error.MobileEquipment.Unknown: Unknown error'
comma@comma-bc60153f:/data/openpilot/system$ mmcli -m 0 --command AT+CFUN=1
response: ''
comma@comma-bc60153f:/data/openpilot/system$ mmcli -m 0 --command AT+QLTS=1
response: '+QLTS: "2023/12/31,23:23:46-20,0"'
```

#### AT+COPS
```
comma@comma-bc60153f:/data/openpilot/system$ mmcli -m 0 --command AT+COPS=2
response: ''
comma@comma-bc60153f:/data/openpilot/system$ mmcli -m 0 --command AT+QLTS=1
response: ''
comma@comma-bc60153f:/data/openpilot/system$ mmcli -m 0 --command AT+QLTS=1
response: '+QIND: SMS DONE'
comma@comma-bc60153f:/data/openpilot/system$ mmcli -m 0 --command AT+QLTS=1
response: '+QLTS: "2023/12/31,23:37:53-20,0"'
```

This one seems to work
#### disable/enable

```
comma@comma-bc60153f:/data/openpilot$ mmcli -m 0 -d
successfully disabled the modem
comma@comma-bc60153f:/data/openpilot$ mmcli -m 0 --command AT+QLTS=1
response: '+QLTS: "2024/01/01,04:00:26-20,0"'
comma@comma-bc60153f:/data/openpilot$ mmcli -m 0 -e
successfully enabled the modem
comma@comma-bc60153f:/data/openpilot$ mmcli -m 0 --command AT+QLTS=1
response: '+QLTS: "2024/01/01,04:00:32-20,0"'
```

**Comment by adeebshihadeh** [comment]
If it’s not straightforward to test like this, then let’s just do active and missing in the test for now, and I’ll manually check inactive before merging. I wouldn’t expect inactive to break if missing/active still work anyway.

**Comment by royjr** [comment]
The `AT+QLTS=1` command doesn't seem to work when physically removing the sim.  Looks like this may be active sim dependant.

> with sim

```
comma@comma-bc60153f:/data/openpilot$ mmcli -m 0 --command AT+QLTS=1
response: '+QLTS: "2024/01/02,16:38:20-20,0"'
```

> with sim removed

```
comma@comma-bc60153f:/data/openpilot$ mmcli -m 0 --command AT+QLTS=1
error: command failed: 'GDBus.Error:org.freedesktop.ModemManager1.Error.MobileEquipment.Unknown: Unknown error'
```

**Comment by adeebshihadeh** [comment]
Yea, for NITZ it has to be registered on the network. I'll move this to a draft for now.

**Comment by adeebshihadeh** [comment]
I looked into it, and I don't think the original bounty target of getting time + timezone in all the specified states is really possible.  Given that, let's adjust the bounty.

We mainly care about time and timezone for two use cases: nav ETA and route names. Route naming is sort of "best effort", so the important one here is nav. Nav only works with GPS anyway, so we can rely on it for getting the time and timezone. We'll also keep your NITZ implementation to speedup some cases where you don't go onroad for a bit. Let's merge `timezoned` into your `timed`, then `timed` will get timezone from GPS coords, timezone from NITZ, and time from NITZ.

**Comment by royjr** [comment]
Correctly failing over to GPS location when available and back on modem when car is turned off. Tests still pending.

```
Setting timezone based on modem
Setting timezone to Etc/GMT+5
Setting time to 2024-01-11 04:42:48
Thu 11 Jan 2024 04:42:48 AM UTC

Setting timezone based on modem
Setting timezone to Etc/GMT+5
Setting time to 2024-01-11 04:42:51
Thu 11 Jan 2024 04:42:51 AM UTC

Setting timezone based on GPS location
Setting timezone to America/New_York
Setting time to @1704948172
Thu 11 Jan 2024 04:42:52 AM UTC

Setting timezone based on GPS location
Setting timezone to America/New_York
Setting time to @1704948174
Thu 11 Jan 2024 04:42:54 AM UTC
```

**Comment by adeebshihadeh** [comment]
Gonna give half the bounty value for this work ($150) since it's not finished and there's been no progress for a couple weeks. I'll take it over from here.

**Pull request closed without merging**